File: src/utils/agentIndex.ts
/**
 * Agent Index: a text-first reformatting of video structure.
 * Agents "read" the film (timeline architecture), not sample arbitrary frames.
 */

export type AgentVideoIndex = {
  policy: 'agent_video_index_v1';
  source: string;
  duration_ms?: number;
  outline: string;
  scene_count: number;
  keyframe_count: number;
  subtitle_cue_count: number;
  transcript_segment_count: number;
  audio_stream_count: number;
  video_stream_count: number;
};

export function buildAgentVideoIndex(doc: {
  provenance: { source: string };
  format?: { duration_ms?: number; format_name?: string };
  streams?: Array<{ codec_type?: string; codec_name?: string; width?: number; height?: number }>;
  scenes?: Array<{ time_ms?: number }>;
  keyframes?: Array<{ time_ms?: number; frame_hash?: string }>;
  subtitles?: Array<{ cues?: unknown[]; text?: string }>;
  transcript?: Array<{ text?: string; start_ms?: number }>;
  warnings?: string[];
}): AgentVideoIndex {
  const streams = doc.streams ?? [];
  const audio = streams.filter((s) => s.codec_type === 'audio');
  const video = streams.filter((s) => s.codec_type === 'video');
  const scenes = doc.scenes ?? [];
  const keyframes = doc.keyframes ?? [];
  const subtitleCueCount = (doc.subtitles ?? []).reduce((n, s) => {
    if (Array.isArray(s.cues)) return n + s.cues.length;
    return n + (s.text ? 1 : 0);
  }, 0);
  const transcript = doc.transcript ?? [];
  const duration = doc.format?.duration_ms;

  const lines: string[] = [
    `# Video index: ${doc.provenance.source}`,
    `- duration_ms: ${duration ?? 'unknown'}`,
    `- container: ${doc.format?.format_name ?? 'unknown'}`,
    `- video_streams: ${video.length}${video[0] ? ` (${video[0].width ?? '?'}x${video[0].height ?? '?'} ${video[0].codec_name ?? ''})` : ''}`,
    `- audio_streams: ${audio.length}${audio[0] ? ` (${audio[0].codec_name ?? ''})` : ''}`,
    `- scenes: ${scenes.length}`,
    `- structural_keyframes: ${keyframes.length}`,
    `- subtitle_cues: ${subtitleCueCount}`,
    `- transcript_segments: ${transcript.length}`,
  ];

  if (scenes.length > 0) {
    lines.push('', '## Scene architecture (not fixed-interval sampling)');
    scenes.slice(0, 40).forEach((s, i) => {
      lines.push(`- scene cut ${i + 1}: ${s.time_ms ?? '?'}ms`);
    });
    if (scenes.length > 40) lines.push(`- … ${scenes.length - 40} more scenes`);
  }

  if (keyframes.length > 0) {
    lines.push('', '## Structural keyframe locators');
    keyframes.slice(0, 32).forEach((k, i) => {
      lines.push(
        `- kf ${i + 1}: t=${k.time_ms ?? '?'}ms hash=${String(k.frame_hash ?? '').slice(0, 12) || 'n/a'}`
      );
    });
  }

  if (subtitleCueCount > 0) {
    lines.push('', '## Subtitles present (embedded) — prefer as dialogue evidence over vision');
  }
  if (transcript.length > 0) {
    lines.push('', '## Transcript segments (optional ASR)');
    transcript.slice(0, 12).forEach((tseg) => {
      const text = String(tseg.text ?? '')
        .replace(/\s+/g, ' ')
        .slice(0, 120);
      lines.push(`- ${tseg.start_ms ?? '?'}ms: ${text}`);
    });
  }

  if ((doc.warnings ?? []).length > 0) {
    lines.push('', '## Warnings', ...(doc.warnings ?? []).map((w) => `- ${w}`));
  }

  lines.push(
    '',
    '## Agent guidance',
    '- Use scenes + keyframe locators for structure; do not invent uniform N-second sampling as architecture.',
    '- For pixel-level claims, follow up with video_evidence render_frame/crop_frame at a locator time_ms.',
    '- Optional LLM vision is non-authority; timeline evidence is the local-first truth.'
  );

  return {
    policy: 'agent_video_index_v1',
    source: doc.provenance.source,
    ...(duration !== undefined ? { duration_ms: duration } : {}),
    outline: lines.join('\n'),
    scene_count: scenes.length,
    keyframe_count: keyframes.length,
    subtitle_cue_count: subtitleCueCount,
    transcript_segment_count: transcript.length,
    audio_stream_count: audio.length,
    video_stream_count: video.length,
  };
}

File: crates/video-reader-core/src/agent_index.rs
//! Text-first timeline outline for agents that cannot inspect pixels directly.

use serde::Serialize;

use crate::frames::KeyframeEvidence;
use crate::scenes::SceneInfo;
use crate::subtitles::SubtitleCue;
use crate::timeline::{ChapterInfo, FormatInfo, StreamInfo};

pub const AGENT_INDEX_POLICY: &str = "agent_video_index_v1";

#[derive(Debug, Clone, Serialize)]
pub struct AgentVideoIndex {
    pub policy: &'static str,
    pub source: String,
    pub duration_ms: u64,
    pub outline: String,
    pub scene_count: u32,
    pub keyframe_count: u32,
    pub subtitle_cue_count: u32,
    pub transcript_segment_count: u32,
    pub audio_stream_count: u32,
    pub video_stream_count: u32,
}

pub fn build_agent_video_index(
    source: &str,
    format: &FormatInfo,
    streams: &[StreamInfo],
    chapters: &[ChapterInfo],
    scenes: &[SceneInfo],
    subtitles: &[SubtitleCue],
    keyframes: &[KeyframeEvidence],
    transcript_segment_count: usize,
    warnings: &[String],
) -> AgentVideoIndex {
    let audio_streams = streams
        .iter()
        .filter(|stream| stream.codec_type == "audio")
        .collect::<Vec<_>>();
    let video_streams = streams
        .iter()
        .filter(|stream| stream.codec_type == "video")
        .collect::<Vec<_>>();

    let mut lines = vec![
        format!("# Video index: {source}"),
        format!("- duration_ms: {}", format.duration_ms),
        format!(
            "- container: {}",
            format.format_name.as_deref().unwrap_or("unknown")
        ),
        format!("- video_streams: {}", video_streams.len()),
        format!("- audio_streams: {}", audio_streams.len()),
        format!("- chapters: {}", chapters.len()),
        format!("- scenes: {}", scenes.len()),
        format!("- timestamp_keyframes: {}", keyframes.len()),
        format!("- subtitle_cues: {}", subtitles.len()),
        format!("- transcript_segments: {transcript_segment_count}"),
    ];

    if !chapters.is_empty() {
        lines.extend(["".into(), "## Chapter segments".into()]);
        for chapter in chapters.iter().take(40) {
            lines.push(format!(
                "- {}ms..{}ms: {}",
                chapter.start_ms,
                chapter.end_ms,
                chapter.title.as_deref().unwrap_or("untitled")
            ));
        }
        if chapters.len() > 40 {
            lines.push(format!("- ... {} more chapters", chapters.len() - 40));
        }
    }

    if !scenes.is_empty() {
        lines.extend(["".into(), "## Scene cut timestamps".into()]);
        for scene in scenes.iter().take(40) {
            lines.push(format!(
                "- scene cut {}: {}ms",
                scene.index + 1,
                scene.time_ms
            ));
        }
        if scenes.len() > 40 {
            lines.push(format!("- ... {} more scene cuts", scenes.len() - 40));
        }
    }

    if !subtitles.is_empty() {
        lines.extend(["".into(), "## Embedded subtitle cues".into()]);
        for cue in subtitles.iter().take(24) {
            lines.push(format!(
                "- {}ms..{}ms: {}",
                cue.start_ms, cue.end_ms, cue.text
            ));
        }
        if subtitles.len() > 24 {
            lines.push(format!("- ... {} more subtitle cues", subtitles.len() - 24));
        }
    }

    if !keyframes.is_empty() {
        lines.extend(["".into(), "## I-frame timestamp locators".into()]);
        for keyframe in keyframes.iter().take(32) {
            lines.push(format!(
                "- kf {}: t={}ms{}",
                keyframe.index + 1,
                keyframe.time_ms,
                keyframe
                    .frame_hash
                    .as_deref()
                    .map(|hash| { format!(" hash={}", hash.chars().take(12).collect::<String>()) })
                    .unwrap_or_default()
            ));
        }
        if keyframes.len() > 32 {
            lines.push(format!("- ... {} more keyframes", keyframes.len() - 32));
        }
    }

    if !warnings.is_empty() {
        lines.extend(["".into(), "## Warnings".into()]);
        lines.extend(warnings.iter().map(|warning| format!("- {warning}")));
    }

    lines.extend([
        "".into(),
        "## Agent guidance".into(),
        "- Use chapter, scene, subtitle, and keyframe timestamps as locators; do not invent a fixed sampling grid.".into(),
        "- Follow up with video_evidence at a locator time_ms for pixel claims.".into(),
        "- Optional ASR or vision remains non-authoritative unless explicitly configured and evidenced.".into(),
    ]);

    AgentVideoIndex {
        policy: AGENT_INDEX_POLICY,
        source: source.into(),
        duration_ms: format.duration_ms,
        outline: lines.join("\n"),
        scene_count: scenes.len() as u32,
        keyframe_count: keyframes.len() as u32,
        subtitle_cue_count: subtitles.len() as u32,
        transcript_segment_count: transcript_segment_count as u32,
        audio_stream_count: audio_streams.len() as u32,
        video_stream_count: video_streams.len() as u32,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::frames::{KeyframeProvenance, KEYFRAME_ROUTE};
    use crate::scenes::SceneProvenance;
    use crate::subtitles::{SubtitleCue, SubtitleProvenance};
    use serde_json::json;

    #[test]
    fn builds_repeatable_outline_with_timestamped_segments() {
        let format = FormatInfo {
            format_name: Some("mp4".into()),
            duration_ms: 5000,
            bit_rate: None,
            size_bytes: None,
            tags: None,
        };
        let streams = vec![
            StreamInfo {
                index: 0,
                codec_type: "video".into(),
                codec_name: Some("h264".into()),
                language: None,
                channels: None,
                sample_rate: None,
                width: Some(160),
                height: Some(120),
                avg_frame_rate: None,
                r_frame_rate: None,
                bit_rate: None,
                disposition: None,
                tags: None,
            },
            StreamInfo {
                index: 1,
                codec_type: "audio".into(),
                codec_name: Some("aac".into()),
                language: None,
                channels: Some(2),
                sample_rate: Some(48_000),
                width: None,
                height: None,
                avg_frame_rate: None,
                r_frame_rate: None,
                bit_rate: None,
                disposition: None,
                tags: None,
            },
        ];
        let chapters = vec![ChapterInfo {
            id: 1,
            start_ms: 0,
            end_ms: 5000,
            title: Some("Intro".into()),
        }];
        let scenes = vec![SceneInfo {
            index: 0,
            time_ms: 1250,
            provenance: SceneProvenance {
                method: "ffmpeg_scene_filter".into(),
                threshold: 0.4,
            },
        }];
        let subtitles = vec![SubtitleCue {
            index: 0,
            start_ms: 100,
            end_ms: 900,
            text: "Hello".into(),
            stream_index: Some(2),
            language: Some("en".into()),
            provenance: SubtitleProvenance {
                method: "ffmpeg_extract".into(),
                format: "srt".into(),
            },
        }];
        let keyframes = vec![KeyframeEvidence {
            index: 0,
            time_ms: 0,
            provenance: KeyframeProvenance {
                method: "ffmpeg_keyframe_select".into(),
                pict_type: "I".into(),
            },
            route: KEYFRAME_ROUTE.into(),
            frame_hash: Some("abcdef0123456789".into()),
            mime: None,
            width: None,
            height: None,
            image_base64: None,
        }];

        let index = build_agent_video_index(
            "/tmp/demo.mp4",
            &format,
            &streams,
            &chapters,
            &scenes,
            &subtitles,
            &keyframes,
            0,
            &["ffmpeg unavailable".into()],
        );
        assert_eq!(index.policy, AGENT_INDEX_POLICY);
        assert_eq!(index.scene_count, 1);
        assert_eq!(index.subtitle_cue_count, 1);
        assert!(index.outline.contains("0ms..5000ms: Intro"));
        assert!(index.outline.contains("1250ms"));
        assert!(index.outline.contains("100ms..900ms: Hello"));
        assert!(index.outline.contains("video_evidence"));
        let serialized = serde_json::to_value(index).expect("serialize");
        assert_eq!(serialized["duration_ms"], json!(5000));
    }
}
